scripts/hud/synchronizer.js

```diff
@@ -0,0 +1,349 @@
+'use strict';
+
+const BUTTONS = {
+	ATTACK: 1 << 0,
+	JUMP: 1 << 1,
+	DUCK: 1 << 2,
+	FORWARD: 1 << 3,
+	BACK: 1 << 4,
+	USE: 1 << 5,
+	CANCEL: 1 << 6,
+	LEFT: 1 << 7,
+	RIGHT: 1 << 8,
+	MOVELEFT: 1 << 9,
+	MOVERIGHT: 1 << 10,
+	ATTACK2: 1 << 11,
+	SCORE: 1 << 16,
+	SPEED: 1 << 17,
+	WALK: 1 << 18,
+	ZOOM: 1 << 19,
+	LOOKSPIN: 1 << 25,
+	BHOPDISABLED: 1 << 29,
+	PAINT: 1 << 30,
+	STRAFE: 1 << 31
+};
+
+const COLORS = {
+	EXTRA: ['rgba(24, 150, 211, 1)', 'rgba(87, 200, 255, 1)'],
+	PERFECT: ['rgba(87, 200, 255, 1)', 'rgba(113, 240, 255, 1)'],
+	GOOD: ['rgba(21, 152, 86, 1)', 'rgba(122, 238, 122, 1)'],
+	SLOW: ['rgba(255, 238, 0, 1)', 'rgba(255, 255, 152, 1)'],
+	NEUTRAL: ['rgba(255, 255, 255, 1)', 'rgba(178, 178, 178, 1)'],
+	LOSS: ['rgba(220, 116, 13, 1)', 'rgba(255, 178, 106, 1)'],
+	STOP: ['rgba(255, 106, 106, 1)', 'rgba(255, 87, 87, 1)']
+};
+
+class Synchronizer {
+	static segments = [$('#Segment0'), $('#Segment1'), $('#Segment2'), $('#Segment3'), $('#Segment4')];
+
+	static container = $('#Container');
+	static needle = $('#Needle');
+
+	static lastSpeed = 0;
+	static lastAngle = 0;
+	static rad2deg = 180 / Math.PI;
+	static deg2rad = 1 / this.rad2deg;
+	static bIsFirstPanelColored = true; // gets toggled in wrapValueToRange()
+	static maxSegmentWidth = 25; // percentage of total element width
+	static firstPanelWidth = this.maxSegmentWidth;
+
+	static onLoad() {
+		this.setSynchroBufferLength();
+
+		this.bIsDefragging = GAMEMODE_WITH_NULL[GameModeAPI.GetCurrentGameMode()].name === 'Defrag';
```

Comment:
Is anything disabling this in non-bhop modes?

Reply:
Oh nvm, just saw how you're doing it in C++, nice way to handle it!

Reply:
should I remove this scaffolding for potential functionality? It's not really doing anything here, didn't think to remove it until now.

Reply:
Isn't this var doing something? I thought some of the strafe bar logic was using it

Reply:
Right now the element isn't drawn in defrag. The c++ logic could easily be changed though.